---
title: "clusterProfiler-singlecell-gene-sets"
author: "Eileen Owens"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Intro
The purpose of this script is to perform and plot GSEA conducted with the R package clusterProfiler. This analysis evaluates the enrichment of user-supplied differential gene expression data for various user-supplied gene sets.

```{r installation, include=FALSE}
# Load necessary packages
library(BiocManager)
library(DESeq2)
library(clusterProfiler)
library(corrplot)
library(dplyr)
library(pathview)
library(enrichplot)
library(ggplot2)
library(ggnewscale)
library(msigdbr)
library(DOSE)
library(ReactomePA)
library(tidyverse)
library(GSEABase)
library(RColorBrewer)
library(ggridges)
library(readr)
library(knitr)
```

## Import data
### Read in differential gene expression data from DESeq2.
```{r data-input}
# Set working directory:
setwd("~/PTCL/PROJ01_26PTCLs/02_scripts")
# Read in data from deseq2
df = read.csv("../01_input/CD4PTCLvsCD4CTRL_DESeq2res_BlankIDsRemoved.csv", header=TRUE)

```
### Rank differentially expressed genes.
```{r convert-probeid, eval=FALSE, echo=FALSE}
# If there is no column of gene symbols (probe ID only), run this code. Otherwise, exclude this code chunk with eval=FALSE and run the next code chunk instead.

# take the stat column from DESeq2 results for ranking
original_gene_list <- df$stat
# name the vector of probe IDs (ENSEMBL gene IDs)
names(original_gene_list) <- df$X
# omit any NA values
gene_list <- na.omit(original_gene_list)
# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

# Install annotations
## Install and load the canine genome annotations. All available annotations can be found here: http://bioconductor.org/packages/release/BiocViews.html#___OrgDb.
organism = "org.Cf.eg.db" # Set the desired organism.
# BiocManager::install(organism, character.only=TRUE) # First time only, then comment out this line.
library(organism, character.only=TRUE)

symbols <- bitr(names(original_gene_list), fromType="ENSEMBL", toType="SYMBOL", OrgDb=organism)
dedup_symbols <- symbols[!duplicated(symbols[c("ENSEMBL")]),] # Removes duplicate IDs
df2 <- df[df$X %in% dedup_symbols$ENSEMBL,] # Creates a new dataframe, df2, which has only the genes which were successfully mapped using the bitr function.
df2$Y <- dedup_symbols$SYMBOL # Creates a new column in df2 with the corresponding gene symbol
deg_genes_list <- df2$stat # Creates a vector of the stat column from DESeq2 output
names(deg_genes_list) <- df2$Y # Names vector with gene symbols
deg_genes_list <- na.omit(deg_genes_list) # Omits any NA values
deg_genes_list = sort(deg_genes_list, decreasing = TRUE) # Sorts in decreasing order
deg_names <- names(deg_genes_list) # Makes a vector of just the ranked gene names

```

```{r rank-symbols}
# If there is a column of gene symbols in the imported deseq2 data, run this code chunk. Otherwise, exclude with eval=FALSE and run the preceding code chunk.

deg_genes_list <- df$stat # Creates a vector of the stat column from DESeq2 output
names(deg_genes_list) <- df$gene_name # Names vector with gene symbols
deg_genes_list <- na.omit(deg_genes_list) # Omits any NA values
deg_genes_list = sort(deg_genes_list, decreasing = TRUE) # Sorts in decreasing order
deg_names <- names(deg_genes_list) # Makes a vector of just the ranked gene names

```

### Import gmt file(s). Gmt files describe gene sets.
```{r mutliple-gmt-import, eval=FALSE, echo=FALSE}
## Input GMT files for specific gene set(s) of interest.
# For importing and combining multiple GMT files in the same directory. Otherwise, exclude with eval=FALSE and run the next code chunk.

setwd("path to gmt") # set working directory to where cell of origin .gmt files are located
gmtFiles <- list.files(pattern = "\\.gmt", full.names = TRUE) # Get list of all .gmt files in this directory
gmtTables <- lapply(gmtFiles, read.gmt) # Apply the read.gmt function to the list of .gmt files and save as a variable
gmt <- do.call(rbind, cellOriginTables) # Rbind files

setwd("~/PTCL/PROJ01_26PTCLs/02_scripts") # Set working directory back to default for the session.
```

```{r single-gmt-import, warning=FALSE}
## Input GMT files for specific gene sets of interest.
# For single gmt file (if combining multiple gmt files, exclude with eval=FALSE and run the preceding code chunk instead).
gmt <- read.gmt.wp("../GSEA/Input/scdata_gmt.gmt", output="data.frame")
gmt <- gmt[, c("name", "gene")]
```
# Perform GSEA analysis
```{r gsea, message=FALSE, warning=FALSE}
## Perform GSEA with order ranked gene list and TERM2GENE = the merged data frames of our annotated gene sets of interest.
gse <- GSEA(deg_genes_list, # Use the sorted list of gene symbols *and* stat
                   exponent = 1,
                   nPerm = 10000,
                   minGSSize = 1,
                   maxGSSize = 800,
                   pvalueCutoff = 1,
                   pAdjustMethod = "BH",
                   TERM2GENE = gmt,
                   verbose = TRUE,
                   by = "fgsea")
```
### Export results
```{r gsea-export}
gse_df <- as.data.frame(gse)
write.csv(gse_df, file="../03_output/230124_cPgsea_singelcelldata.csv")
```

## Dotplot
```{r dotplot, message=FALSE, warning=FALSE, fig.height=6, fig.width=8}
gse %>%
  dotplot(showCategory = 20, x = "NES") +
  scale_colour_viridis_c(name = "Adjusted\nP-value",
                         option="H") +
  geom_vline(xintercept = 0, linetype=2)
```

## Ridgeplot
```{r ridgeplot, warning=FALSE, message=FALSE, fig.height=6, fig.width=8}
ridgeplot(gse, showCategory=20) + labs(x = "enrichment distribution")
```

## Enrichment plots
```{r get-enriched-paths, eval=FALSE, echo=FALSE}
# This section will not be included in final knit
enrichedPaths <- gse$ID # Create a character vector of all enriched gene sets
enrichedPaths # View the list in the console for easy copy/paste
```

```{r enrichment-plot, echo=FALSE, fig.height=14, fig.width=16, fig.fullwidth=TRUE}
# for geneSetID, either individually list gene sets to plot with c("GENE_SET_1", "GENE_SET_2") or use character vector variable from above to plot all enriched sets.

mypalette4 <- brewer.pal(4, "Set1")[c(1,2,3,4)]
mypalette6 <- brewer.pal(6, "Set1")[c(1,2,3,4,5,7)] # Adjust as needed for the number of gene sets to be mapped on the figure. If using 6, recommended to exclude 6 (yellow).
mypalette5 <- brewer.pal(5, "Set1")[c(1,2,3,4,5)]
mypalette7 <- brewer.pal(7, "Set1")[c(1,2,3,4,5,7,8)]

ep1 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster7",
                        "K9ThymusCluster10",
                        "K9ThymusCluster18",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for Thymic Pre_T_DN cells, BM T-cells, and LN T-cells",
          color = mypalette6
)
ep1

ep2 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster4",
                        "K9ThymusCluster21",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for Thymic ISP_T_Cells, BM T-cells, and LN T-cells",
          color = mypalette5
)
ep2

ep3 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster8",
                        "K9ThymusCluster6",
                        "K9ThymusCluster1",
                        "K9ThymusCluster5",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for Early and Late Thymic DP_T_cells Compared to Bone Marrow and Lymph Node T-cells",
          color = mypalette7
)
ep3

ep4 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster9",
                        "K9ThymusCluster15",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for Dev SP_T_cells Compared to Bone Marrow and Lymph Node T-cells",
          color = mypalette5
)
ep4

ep5 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster19",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for CD4 SP_T_cells Compared to Bone Marrow and Lymph Node T-cells",
          color = mypalette4
)
ep5

ep6 <- gseaplot2(gse, 
          geneSetID = c("K9ThymusCluster14",
                        "K9ThymusCluster3",
                        "K9BoneMarrowCluster21",
                        "K9BoneMarrowCluster26",
                        "K9LymphNodeCluster0"),
          title = "Enrichment for CD8 SP_T_cells and CD8 T-cells Compared to Bone Marrow and Lymph Node T-cells",
          color = mypalette4
)
ep6

```
```{r}
cnet <- cnetplot(gse, categorySize="pvalue", foldChange=gene_list, node_label="all", cex_label_category=0.8, cex_label_gene=0.4, showCategory=24)
```

# Citations
```{r}
sessionInfo()
citation()
```
