---
title: "RNAseq-Analysis"
author: "Eileen Owens"
date: "`r Sys.Date()`"
output:
  html_document:
  keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,
                      cache = TRUE,
                      warning = FALSE)
```
# 1. Loading data and packages

## 1.1 Install necessary software packages (first time only):
```{r installation, results="hide"}
if (!require("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
BiocManager::install(version = "3.16", force=TRUE)

options(BioC_mirror = "http://bioconductor.org")

BiocManager::install("DESeq2")
BiocManager::install("apeglm")
install.packages("corrplot")
install.packages("pheatmap")
install.packages("RColorBrewer")
install.packages("rlang")
BiocManager::install("clusterProfiler")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
BiocManager::install("EnhancedVolcano")
```

## 1.2 Load necessary software packages (every time):
```{r software-load, results="hide", message=FALSE}
library(DESeq2)
library(corrplot)
library(RColorBrewer)
library(pheatmap)
library(apeglm)
library(tidyverse)
library(clusterProfiler)
library(enrichplot)
```

## 1.3 Read in the featureCounts data and metadata:
```{r data-input, results="hide"}
# Set the appropriate working directory.
setwd("/Users/eileen/PTCL/PROJ01_26PTCLs/02_scripts")

# Read in the FeatureCounts file.
countsData <- read.table(file = "../01_input/220628_featureCounts_withCanFam31Alignment.txt", header = FALSE, row.names = 1, skip = 2) # row.names = 1 assigns the first column (geneid) as the row names for the data frame.

# Read in the metadata file. This script assumes that the paired fasta filenames are in the first two columns, the sample ID or nickname is in the third column, and the experiment group or phenotype is in the fourth column.
metadata1 <- read.table(file = "../01_input/2022_metadata.txt", header = FALSE)
colnames(metadata1) <- c("fasta1", "fasta2", "sample_name", "phenotype") # Adds column headers

### Give column names to countsData file: ###
as.vector(metadata1$sample_name) # The last column names will be names for each sample. We can pull those names from metadata1.
colnames(countsData) <- c("chr", "start", "stop", "strand", "length", as.vector(metadata1$sample_name))
```

## 1.4 Make count matrix (cts) file for DESeq2

If using all samples, run the following chunk. Otherwise, comment out.
```{r cts, results="hide"}
head(countsData[,6:45]) # Adjust the indexing as needed to include only the sample columns with read counts. This also keeps the row names (the geneid column)

# Save just the subset as an object called cts:
cts <- as.matrix(countsData[,6:45]) 
```
If using only only CD4 PTCLs and CD4 controls, run this code chunk. Otherwise, comment out.

```{r cts, results="hide"}
# To select for only CD4 PTCLs and CD4 controls, include this code. Otherwise, comment out to include all samples.
colnames(countsData)
countsData_CD4 <- dplyr::select(countsData, -c("chr", "start", "stop", "strand", "length", "CF04", "CF06", "CF08", "CF09", "CF10", "CF11", "CF13", "CF14", "CF16", "CF38")) # Removes all columns except the count data for CD4 PTCLs and CD4 controls.

# Save just the subset as an object called cts:
cts <- as.matrix(countsData_CD4)
```

## 1.5 Make coldata file for DESeq2

If running all samples, use the following code chunk. Otherwise, comment out.
```{r coldata}
# Reorganize the metadata table so the sample_name column values are now row headers.
rownames(metadata1)<- metadata1$sample_name
coldata <- metadata1[, c("phenotype"), drop=FALSE] # drop=FALSE keeps row names
coldata$phenotype <- as.factor(coldata$phenotype)

# Ensure that the rownames of coldata exactly match the colnames of cts:
all(rownames(coldata) == colnames(cts)) # Should return TRUE
```

If using only CD4 PTCLs and CD4 controls, run the following code chunk. Otherwise, comment out.
```{r coldata}
# Reorganize the metadata table so the sample_name column values are now row headers.
metadata_CD4 <- read.csv("../01_input/metadata_CD4sOnly.csv")
colnames(metadata_CD4)[2] <- "phenotype" # make phenotype column name lowercase
rownames(metadata_CD4) <- metadata_CD4$Sample
coldata <- metadata_CD4[, c("phenotype"), drop=FALSE] # drop=FALSE keeps row names
coldata$phenotype <- as.factor(coldata$phenotype)

# Ensure that the rownames of coldata exactly match the colnames of cts:
all(rownames(coldata) == colnames(cts)) # Should return TRUE
```

# 2. Running a differential expression analysis with DESeq2

## 2.1 Set parameters for the study
```{r dds, results="hide"}
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ phenotype)
```

## 2.2 Filter for present genes
Excludes all samples that have less than 10 reads.
```{r filter-DESeq2-output-1, results="hide"}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

## 2.3 Remove any samples to be excluded from the differential expression analysis (previously determined outliers, etc.) 
```{r filter-DESeq2-output-2, results='hide'}
# Comment this section out if no samples are to be excluded from the analysis.
colnames(dds) # See which column contains the outlier; in this case, we found CF21 to be an outlier and wish to remove it.
dds <- dds[,-12]
colnames(dds)
```

## 2.4 Optional: Set factor levels
Since this experiment utilized two control groups (CD4_CTRL and CD8_CTRL), I chose to comment this section out and explicitly tell the "results" function which comparison to make each time.
```{r factor-levels, results="hide"}
# There are 2 ways to set the factor level:
#dds$condition <- factor(dds$phenotype, levels = c("CD4_CTRL","CD4_PTCL","CD8_PTCL")) # Control group is listed first.
#dds$condition <- relevel(dds$phenotype, ref = "CD4_CTRL") # Relevel requires only specifying the reference level.
```

## 2.5A Perform DESeq2 analysis
```{r DESeq2, results="hide", message=FALSE}
dds <- DESeq(dds)
```

## 2.5B Save the dds object for ease of import in any downstream analyses.
```{r RDS, eval=FALSE}
saveRDS(dds, file="filename")
```

## 2.6 Plot dispersion
```{r plot-dispersion, message=FALSE, fig.height=5, fig.width=8, fig.fullwidth=TRUE}
plotDispEsts(dds)
```

## 2.7 Perform a principal component analysis to identify any outliers before proceeding.
Take the r-stabilized log transformations of all the normalized count data and perform the plotPCA function.
```{r PCA, results='hide'}
rld <- rlog(dds, blind=TRUE)
pcaData <- plotPCA(rld, intgroup = "phenotype", returnData = TRUE)
```

### 2.7.1 Export the intgroup covariates used for PCA plotting:
```{r PCA-export}
write.csv(pcaData, file="../03_output/PCA_data.csv")
```

### 2.7.3 Draw the PCA plot
```{r PCA-plot, message=FALSE, fig.height=8, fig.width=6, fig.fullwidth=TRUE}
pcaData <- plotPCA(rld, intgroup = "phenotype", returnData = FALSE) +
  geom_text(aes(label = colnames(rld)), position = position_nudge(y = -2)) # Uses the name column of rld as the dot label.
pcaData
```
If principal component analysis reveals any outliers, go back up to section 2.4 and execute that code.


## 2.8 Differential expression analysis
Calculate the statistically significant differences between conditions. Specify the coefficient or contrast you want to build a results table for; e.g., CD4_PTCL vs. CD8_PTCL, CD4_PTCL vs CD4_CTRL, etc. Save each comparison to its own variable. Follow up with log fold change shrinkage for visualization and gene ranking.

```{r DEG-instructions, include=FALSE}
# The appropriate format is as follows: 
#BvsC <- results(dds, contrast=c("condition", "B", "C"))
#AvsB <- results(dds,contrast=c("condition", "A", "B"))
```

```{r res-file-generation, results='hide', message=FALSE}
res <- results(dds)

# If you set factor levels earlier, uncomment this section; otherwise, run a different "results" function for each comparison.
#resultsNames(dds) # lists the coefficients
#resLFC <- lfcShrink(dds, coef="phenotype_CD4_PTCL_vs_CD4_CTRL", type="apeglm")
#head(resLFC)

CD4PTCLvsCD4CTRL_results <- results(dds, contrast=c("phenotype", "CD4_PTCL", "CD4_CTRL"))
CD4PTCLvsCD4CTRL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD4_PTCL", "CD4_CTRL"), res=CD4PTCLvsCD4CTRL_results, type="normal")

CD4PTCLvsCD8PTCL_results <- results(dds, contrast=c("phenotype", "CD4_PTCL", "CD8_PTCL"))
CD4PTCLvsCD8PTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD4_PTCL", "CD8_PTCL"), res=CD4PTCLvsCD8PTCL_results, type="normal")

CD4PTCLvsDNPTCL_results <- results(dds, contrast=c("phenotype", "CD4_PTCL", "DN_PTCL"))
CD4PTCLvsDNPTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD4_PTCL", "DN_PTCL"), res=CD4PTCLvsDNPTCL_results, type="normal")

CD8PTCLvsCD8CTRL_results <- results(dds, contrast=c("phenotype", "CD8_PTCL", "CD8_CTRL"))
CD8PTCLvsCD8CTRL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD8_PTCL", "CD8_CTRL"), res=CD8PTCLvsCD8CTRL_results, type="normal")

CD8PTCLvsCD4PTCL_results <- results(dds, contrast=c("phenotype", "CD8_PTCL", "CD4_PTCL"))
CD8PTCLvsCD4PTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD8_PTCL", "CD4_PTCL"), res=CD8PTCLvsCD4PTCL_results, type="normal")

CD8PTCLvsDNPTCL_results <- results(dds, contrast=c("phenotype", "CD8_PTCL", "DN_PTCL"))
CD8PTCLvsDNPTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD8_PTCL", "DN_PTCL"), res=CD8PTCLvsDNPTCL_results, type="normal")

DNPTCLvsCD4PTCL_results <- results(dds, contrast=c("phenotype", "DN_PTCL", "CD4_PTCL"))
DNPTCLvsCD4PTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "DN_PTCL", "CD4_PTCL"), res=DNPTCLvsCD4PTCL_results, type="normal")

DNPTCLvsCD8PTCL_results <- results(dds, contrast=c("phenotype", "DN_PTCL", "CD8_PTCL"))
DNPTCLvsCD8PTCL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "DN_PTCL", "CD8_PTCL"), res=DNPTCLvsCD8PTCL_results, type="normal")

DNPTCLvsCD4CTRL_results <- results(dds, contrast=c("phenotype", "DN_PTCL", "CD4_CTRL"))
DNPTCLvsCD4CTRL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "DN_PTCL", "CD4_CTRL"), res=DNPTCLvsCD4CTRL_results, type="normal")

DNPTCLvsCD8CTRL_results <- results(dds, contrast=c("phenotype", "DN_PTCL", "CD8_CTRL"))
DNPTCLvsCD8CTRL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "DN_PTCL", "CD8_CTRL"), res=DNPTCLvsCD8CTRL_results, type="normal")

CD4PTCLvsCD8CTRL_results <- results(dds, contrast=c("phenotype", "CD4_PTCL", "CD8_CTRL"))
CD4PTCLvsCD8CTRL_shrink <- lfcShrink(dds=dds, contrast=c("phenotype", "CD4_PTCL", "CD8_CTRL"), res=CD4PTCLvsCD8CTRL_results, type="normal")
```

# 3. Exporting DESeq2 Data
## 3.1 Exporting results tables
Add a column with gene names, as the results tables in their current form only have the probe ID.
```{r DESeq2-data-export, results='hide'}
# Import a csv file where the first column is labeled "probe_id" and contains the probe IDs, and the second column is labeled "gene_name" and contains the preferred gene names.
genenames <- read.csv('../01_input/gene.description.csv')

# Add a column to the results table containing the probe IDs, which are currently the rownames.
CD4PTCLvsCD4CTRL_shrink$probe_id <- rownames(CD4PTCLvsCD4CTRL_shrink)

# Merge the genenames metadata file with the results tables by probe ID.
CD4PTCLvsCD4CTRL_final_res <- merge(as(CD4PTCLvsCD4CTRL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE) # The all.x=TRUE argument says that it should include rows of CD4PTCLvsCD4CTRL_shrink even if there is no matching row of the genenames table.

# Convert probe ID back the rownames and remove the probe ID column.
rownames(CD4PTCLvsCD4CTRL_final_res) <- CD4PTCLvsCD4CTRL_final_res$probe_id
CD4PTCLvsCD4CTRL_final_res <- CD4PTCLvsCD4CTRL_final_res[,-1]

# Export the final result as a .csv file.
write.csv(CD4PTCLvsCD4CTRL_final_res, file="../03_output/CD4PTCLvsCD4CTRL_DESeq2res.csv")
```

Repeat for all comparisons.
```{r DESeq2-data-export-2, results='hide'}
CD4PTCLvsCD8PTCL_shrink$probe_id <- rownames(CD4PTCLvsCD8PTCL_shrink)
CD4PTCLvsCD8PTCL_final_res <- merge(as(CD4PTCLvsCD8PTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD4PTCLvsCD8PTCL_final_res) <- CD4PTCLvsCD8PTCL_final_res$probe_id
CD4PTCLvsCD8PTCL_final_res <- CD4PTCLvsCD8PTCL_final_res[,-1]
write.csv(CD4PTCLvsCD8PTCL_final_res, file="../03_output/CD4PTCLvsCD8PTCL_DESeq2res.csv")

CD4PTCLvsDNPTCL_shrink$probe_id <- rownames(CD4PTCLvsDNPTCL_shrink)
CD4PTCLvsDNPTCL_final_res <- merge(as(CD4PTCLvsDNPTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD4PTCLvsDNPTCL_final_res) <- CD4PTCLvsDNPTCL_final_res$probe_id
CD4PTCLvsDNPTCL_final_res <- CD4PTCLvsDNPTCL_final_res[,-1]
write.csv(CD4PTCLvsDNPTCL_final_res, file="../03_output/CD4PTCLvsDNPTCL_DESeq2res.csv")

CD4PTCLvsCD8CTRL_shrink$probe_id <- rownames(CD4PTCLvsCD8CTRL_shrink)
CD4PTCLvsCD8CTRL_final_res <- merge(as(CD4PTCLvsCD8CTRL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD4PTCLvsCD8CTRL_final_res) <- CD4PTCLvsCD8CTRL_final_res$probe_id
CD4PTCLvsCD8CTRL_final_res <- CD4PTCLvsCD8CTRL_final_res[,-1]
write.csv(CD4PTCLvsCD8CTRL_final_res, file="../03_output/CD4PTCLvsCD8CTRL_DESeq2res.csv")

CD8PTCLvsCD8CTRL_shrink$probe_id <- rownames(CD8PTCLvsCD8CTRL_shrink)
CD8PTCLvsCD8CTRL_final_res <- merge(as(CD8PTCLvsCD8CTRL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD8PTCLvsCD8CTRL_final_res) <- CD8PTCLvsCD8CTRL_final_res$probe_id
CD8PTCLvsCD8CTRL_final_res <- CD8PTCLvsCD8CTRL_final_res[,-1]
write.csv(CD8PTCLvsCD8CTRL_final_res, file="../03_output/CD8PTCLvsCD8CTRL_DESeq2res.csv")

CD8PTCLvsCD4PTCL_shrink$probe_id <- rownames(CD8PTCLvsCD4PTCL_shrink)
CD8PTCLvsCD4PTCL_final_res <- merge(as(CD8PTCLvsCD4PTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD8PTCLvsCD4PTCL_final_res) <- CD8PTCLvsCD4PTCL_final_res$probe_id
CD8PTCLvsCD4PTCL_final_res <- CD8PTCLvsCD4PTCL_final_res[,-1]
write.csv(CD8PTCLvsCD4PTCL_final_res, file="../03_output/CD8PTCLvsCD4PTCL_DESeq2res.csv")

CD8PTCLvsDNPTCL_shrink$probe_id <- rownames(CD8PTCLvsDNPTCL_shrink)
CD8PTCLvsDNPTCL_final_res <- merge(as(CD8PTCLvsDNPTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(CD8PTCLvsDNPTCL_final_res) <- CD8PTCLvsDNPTCL_final_res$probe_id
CD8PTCLvsDNPTCL_final_res <- CD8PTCLvsDNPTCL_final_res[,-1]
write.csv(CD8PTCLvsDNPTCL_final_res, file="../03_output/CD8PTCLvsDNPTCL_DESeq2res.csv")

DNPTCLvsCD4PTCL_shrink$probe_id <- rownames(DNPTCLvsCD4PTCL_shrink)
DNPTCLvsCD4PTCL_final_res <- merge(as(DNPTCLvsCD4PTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(DNPTCLvsCD4PTCL_final_res) <- DNPTCLvsCD4PTCL_final_res$probe_id
DNPTCLvsCD4PTCL_final_res <- DNPTCLvsCD4PTCL_final_res[,-1]
write.csv(DNPTCLvsCD4PTCL_final_res, file="../03_output/DNPTCLvsCD4PTCL_DESeq2res.csv")

DNPTCLvsCD8PTCL_shrink$probe_id <- rownames(DNPTCLvsCD8PTCL_shrink)
DNPTCLvsCD8PTCL_final_res <- merge(as(DNPTCLvsCD8PTCL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(DNPTCLvsCD8PTCL_final_res) <- DNPTCLvsCD8PTCL_final_res$probe_id
DNPTCLvsCD8PTCL_final_res <- DNPTCLvsCD8PTCL_final_res[,-1]
write.csv(DNPTCLvsCD8PTCL_final_res, file="../03_output/DNPTCLvsCD8PTCL_DESeq2res.csv")

DNPTCLvsCD4CTRL_shrink$probe_id <- rownames(DNPTCLvsCD4CTRL_shrink)
DNPTCLvsCD4CTRL_final_res <- merge(as(DNPTCLvsCD4CTRL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(DNPTCLvsCD4CTRL_final_res) <- DNPTCLvsCD4CTRL_final_res$probe_id
DNPTCLvsCD4CTRL_final_res <- DNPTCLvsCD4CTRL_final_res[,-1]
write.csv(DNPTCLvsCD4CTRL_final_res, file="../03_output/DNPTCLvsCD4CTRL_DESeq2res.csv")

DNPTCLvsCD8CTRL_shrink$probe_id <- rownames(DNPTCLvsCD8CTRL_shrink)
DNPTCLvsCD8CTRL_final_res <- merge(as(DNPTCLvsCD8CTRL_shrink,"data.frame"), genenames, by="probe_id", all.x=TRUE)
rownames(DNPTCLvsCD8CTRL_final_res) <- DNPTCLvsCD8CTRL_final_res$probe_id
DNPTCLvsCD8CTRL_final_res <- DNPTCLvsCD8CTRL_final_res[,-1]
write.csv(DNPTCLvsCD8CTRL_final_res, file="../03_output/DNPTCLvsCD8CTRL_DESeq2res.csv")
```

## 3.2 Exporting normalized count data
```{r export-normalized-counts, results='hide'}
sizeFactors(dds)
normalized_counts <- as.data.frame(counts(dds, normalized=TRUE))
normalized_counts$probe_id <- rownames(normalized_counts) # Add a column with probe ID.
normalized_counts <- merge(as(normalized_counts,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
write.csv(normalized_counts, file="../03_output/NormalizedCounts.csv")
```

## 3.3 Exporting lists of differentially expressed genes.
```{r export-DEG, results='hide'}
# Select the significant subset of genes that are up-regulated:
Up_in_CD4PTCL <- subset(CD4PTCLvsCD4CTRL_shrink, padj < 0.05 & log2FoldChange > 0.5)
Up_in_CD4PTCL <- Up_in_CD4PTCL[order(Up_in_CD4PTCL$padj),] #order them

# Select the significant subset of genes that are down-regulated
Down_in_CD4PTCL <- subset(CD4PTCLvsCD4CTRL_shrink, padj < 0.05 & log2FoldChange < -0.5)
Down_in_CD4PTCL <- Down_in_CD4PTCL[order(Down_in_CD4PTCL$padj),]

# Save these lists to output files:
write(rownames(Up_in_CD4PTCL), file = "../03_output/CD4PTCL_vs_CD4CTRL_DEG_UP.txt", sep = "\n")
write(rownames(Down_in_CD4PTCL), file = "../03_output/CD4PTCL_vs_CD4CTRL_DEG_DOWN.txt", sep = "\n")
write(rownames(CD4PTCLvsCD4CTRL_shrink), file = "../03_output/all_present_genes_CD4PTCLvsCD4CTRL.txt", sep = "\n") # for a control/background set of genes
```
With these lists of differentially expressed genes, we can see what KEGG pathways or GO Ontology terms are associated: https://david.ncifrf.gov/

# 4. Data Visualization

## 4.1 MA plots
Shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet.

```{r MAplot, fig.height=6, fig.width=10, fig.fullwidth=TRUE}
plotMA(CD4PTCLvsCD4CTRL_shrink, main="CD4 PTCL vs CD4 Control", ylim = c(-7,7), 
       ylab = "log fold change",
       xlab = "means of normalized counts")
```

## 4.2 Plot Counts
A function for examining the counts of reads for a single gene across the groups. It normalizes counts by the estimated size factors (or normalization factors if these were used) and adds a pseudocount of 1/2 to allow for log scale plotting. The counts are grouped by the variables in "intgroup", where more than one variable can be specified.
```{r PlotCounts, message=FALSE}
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000005019"), intgroup="phenotype") #GATA3
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000016705"), intgroup="phenotype") # TBX21
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000015670"), intgroup="phenotype") # PTEN
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000011722"), intgroup="phenotype") # CD34
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000002065"), intgroup="phenotype") # CD117
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000008507"), intgroup="phenotype") # DNTT
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000000812"), intgroup="phenotype") # DLA-DQA1 (Dog MHC II)
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000000848"), intgroup="phenotype") # DLA-DMA (Dog MHC II)
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000000896"), intgroup="phenotype") # DLA-DOA (Dog MHC II)
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000018931"), intgroup="phenotype") # MHC class II transactivator CIITA
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000010955"), intgroup="phenotype") # PTHLH
plotCounts(dds, gene=which(rownames(res) == "ENSCAFG00000005210"), intgroup="phenotype") # IL2RA (CD25)
```

## 4.3 Plot multiple genes with phenotype labels
Make the lists of genes to include in each plot.
```{r gene-lists, results='hide'}
# T cell phenotypes
CD8genes <- c("EOMES", "IFNG", "TNF", "PRF1", "GZMB", "GZMA")
Th1genes <- c("TBX21", "IFNG", "TNF")
Th2genes <- c("GATA3", "IL4R", "IL5", "IL13")
Th17genes <- c("RORC", "STAT3", "IL17A", "IL17C", "IL17F")
Treggenes <- c("FOXP3", "IL10", "TGFB1")
Tfhgenes <- c("BCL6", "IL21R", "CXCL13", "CXCR5")

# TCR signaling genes
TCRgenes <- c("CD3E", "CD3D", "CD3G", "CD28", "ZAP70", "LCK", "FYN", "IL2RA", "IL2RB", "IL2RG")

# Markers of immaturity
precursorGenes <- c("CD34", "KIT", "DNTT")
```

Subset the normalized counts matrix (the version that includes a column of gene names) for each group, then gather the columns to get the normalized counts in a single column.
```{r gene-list-subsets, results='hide', message=FALSE}
CD8subset <- normalized_counts %>%
  filter(gene_name %in% CD8genes)
gathered_CD8subset <- CD8subset %>%
  gather(colnames(CD8subset)[2:41], key = "sample_name", value="normalized_counts")
gathered_CD8subset$normalized_counts <- as.numeric(as.character(gathered_CD8subset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.
```

Repeat for all subsets.
```{r gene-list-subsets-2, results='hide', message=FALSE}
Th1subset <- normalized_counts %>%
  filter(gene_name %in% Th1genes)
Th2subset <- normalized_counts %>%
  filter(gene_name %in% Th2genes)
Th17subset <- normalized_counts %>%
  filter(gene_name %in% Th17genes)
Tregsubset <- normalized_counts %>%
  filter(gene_name %in% Treggenes)
Tfhsubset <- normalized_counts %>%
  filter(gene_name %in% Tfhgenes)
TCRsubset <- normalized_counts %>%
  filter(gene_name %in% TCRgenes)
precursorSubset <- normalized_counts %>%
  filter(gene_name %in% precursorGenes)

gathered_Th1subset <- Th1subset %>%
  gather(colnames(Th1subset)[2:41], key = "sample_name", value="normalized_counts")
gathered_Th1subset$normalized_counts <- as.numeric(as.character(gathered_Th1subset$normalized_counts))

gathered_Th2subset <- Th2subset %>%
  gather(colnames(Th2subset)[2:41], key = "sample_name", value="normalized_counts")
gathered_Th2subset$normalized_counts <- as.numeric(as.character(gathered_Th2subset$normalized_counts))

gathered_Th17subset <- Th17subset %>%
  gather(colnames(Th17subset)[2:41], key = "sample_name", value="normalized_counts")
gathered_Th17subset$normalized_counts <- as.numeric(as.character(gathered_Th17subset$normalized_counts))

gathered_Tregsubset <- Tregsubset %>%
  gather(colnames(Tregsubset)[2:41], key = "sample_name", value="normalized_counts")
gathered_Tregsubset$normalized_counts <- as.numeric(as.character(gathered_Tregsubset$normalized_counts))

gathered_Tfhsubset <- Tfhsubset %>%
  gather(colnames(Tfhsubset)[2:41], key = "sample_name", value="normalized_counts")
gathered_Tfhsubset$normalized_counts <- as.numeric(as.character(gathered_Tfhsubset$normalized_counts))

gathered_TCRsubset <- TCRsubset %>%
  gather(colnames(TCRsubset)[2:41], key = "sample_name", value="normalized_counts")
gathered_TCRsubset$normalized_counts <- as.numeric(as.character(gathered_TCRsubset$normalized_counts))

gathered_precursorSubset <- precursorSubset %>%
  gather(colnames(precursorSubset)[2:41], key = "sample_name", value="normalized_counts")
gathered_precursorSubset$normalized_counts <- as.numeric(as.character(gathered_precursorSubset$normalized_counts))

```

Combine with the metadata information to allow coloring of counts by sample group.
```{r gene-list-subsets-3, results='hide', message=FALSE}
metadata2 <- metadata1[,3:4] # Reduce the metadata table to just the sample_name and phenotype columns.
gathered_CD8subset <- inner_join(metadata2, gathered_CD8subset) # This will merge the 2 data frames with respect to the "sample_name" column (i.e., a column with the same column name in both data frames)
gathered_CD8subset <- merge(as(gathered_CD8subset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
```

Repeat for all groups.
```{r gene-list-subsets-4, results='hide', message=FALSE}
gathered_Th1subset <- inner_join(metadata2, gathered_Th1subset)
gathered_Th1subset <- merge(as(gathered_Th1subset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_Th2subset <- inner_join(metadata2, gathered_Th2subset)
gathered_Th2subset <- merge(as(gathered_Th2subset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_Th17subset <- inner_join(metadata2, gathered_Th17subset)
gathered_Th17subset <- merge(as(gathered_Th17subset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_Tregsubset <- inner_join(metadata2, gathered_Tregsubset)
gathered_Tregsubset <- merge(as(gathered_Tregsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_Tfhsubset <- inner_join(metadata2, gathered_Tfhsubset)
gathered_Tfhsubset <- merge(as(gathered_Tfhsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_TCRsubset <- inner_join(metadata2, gathered_TCRsubset)
gathered_TCRsubset <- merge(as(gathered_TCRsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE)

gathered_precursorSubset <- inner_join(metadata2, gathered_precursorSubset)
gathered_precursorSubset <- merge(as(gathered_precursorSubset,"data.frame"), genenames, by="probe_id", all.x=TRUE)
```

Plot the counts.
```{r plot-counts-gene-lists, message=FALSE}
jitter <- position_jitter(width=0.1, height=0.1)

ggplot(gathered_CD8subset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of CD8 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Th1subset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th1 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Th2subset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th2 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Th17subset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th17 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Tregsubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Treg Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Tfhsubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Tfh Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_TCRsubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Genes Associated with TCR Signaling") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_precursorSubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Markers of Immaturity") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))
```

## 4.4 Plot counts for CD4 PTCL vs CD4 Controls only.
```{r plot-counts-CD4s, message=FALSE}
CD4samples <- c("probe_id", "CF01", "CF02", "CF03", "CF05", "CF07", "CF12", "CF15", "CF17", "CF18", "CF19", "CF20", "CF22", "CF23", "CF24", "CF25", "CF26", "CF27", "CF28", "CF29", "CF30", "CF31", "CF32", "CF33", "CF34", "CF35", "CF36", "CF37", "CF39", "CF40", "gene_name")
normalized_counts_CD4 <- normalized_counts[, CD4samples]

precursorSubset <- normalized_counts_CD4 %>%
  filter(gene_name %in% precursorGenes)
gathered_precursorSubset <- precursorSubset %>%
  gather(colnames(precursorSubset)[2:31], key = "sample_name", value="normalized_counts")
gathered_precursorSubset <- inner_join(metadata2, gathered_precursorSubset) # Not sure if this will work with the additional phenotypes still in metadata2; may have to cut down the metadata to just CD4s.
gathered_precursorSubset <- merge(as(gathered_precursorSubset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_precursorSubset$normalized_counts <- as.numeric(as.character(gathered_precursorSubset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

PTCLsubgroupGenes <- c("GATA3", "TBX21")
PTCLsubgroups <- normalized_counts_CD4 %>%
  filter(gene_name %in% PTCLsubgroupGenes)
gathered_PTCLsubgroups <- PTCLsubgroups %>%
  gather(colnames(PTCLsubgroups)[2:31], key = "sample_name", value="normalized_counts")
gathered_PTCLsubgroups <- inner_join(metadata2, gathered_PTCLsubgroups)
gathered_PTCLsubgroups <- merge(as(gathered_PTCLsubgroups,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_PTCLsubgroups$normalized_counts <- as.numeric(as.character(gathered_PTCLsubgroups$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

Activatedgenes <- c("IL2RA", "DLA-DRA",  "DLA-DQA1", "HLA-DQB1", "HLA-DQB2")
Activatedsubset <- normalized_counts_CD4 %>%
  filter(gene_name %in% Activatedgenes)
gathered_Activatedsubset <- Activatedsubset %>%
  gather(colnames(Activatedsubset)[2:31], key = "sample_name", value="normalized_counts")
gathered_Activatedsubset <- inner_join(metadata2, gathered_Activatedsubset)
gathered_Activatedsubset <- merge(as(gathered_Activatedsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_Activatedsubset$normalized_counts <- as.numeric(as.character(gathered_Activatedsubset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

CD4TCRsubset <- normalized_counts_CD4 %>%
  filter(gene_name %in% TCRgenes)
gathered_CD4TCRsubset <- CD4TCRsubset %>%
  gather(colnames(PTCLsubgroups)[2:31], key = "sample_name", value="normalized_counts")
gathered_CD4TCRsubset <- inner_join(metadata2, gathered_CD4TCRsubset)
gathered_CD4TCRsubset <- merge(as(gathered_CD4TCRsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_CD4TCRsubset$normalized_counts <- as.numeric(as.character(gathered_CD4TCRsubset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

CD4Th1subset <- normalized_counts_CD4 %>%
  filter(gene_name %in% Th1genes)
gathered_CD4Th1subset <- CD4Th1subset %>%
  gather(colnames(CD4Th1subset)[2:31], key = "sample_name", value="normalized_counts")
gathered_CD4Th1subset <- inner_join(metadata2, gathered_CD4Th1subset)
gathered_CD4Th1subset <- merge(as(gathered_CD4Th1subset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_CD4Th1subset$normalized_counts <- as.numeric(as.character(gathered_CD4Th1subset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

CD4Th2subset <- normalized_counts_CD4 %>%
  filter(gene_name %in% Th2genes)
gathered_CD4Th2subset <- CD4Th2subset %>%
  gather(colnames(CD4Th2subset)[2:31], key = "sample_name", value="normalized_counts")
gathered_CD4Th2subset <- inner_join(metadata2, gathered_CD4Th2subset)
gathered_CD4Th2subset <- merge(as(gathered_CD4Th2subset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
gathered_CD4Th2subset$normalized_counts <- as.numeric(as.character(gathered_CD4Th2subset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

CD4Th17subset <- normalized_counts_CD4 %>%
  filter(gene_name %in% Th17genes)
gathered_CD4Th17subset <- CD4Th17subset %>%
  gather(colnames(CD4Th17subset)[2:31], key = "sample_name", value="normalized_counts")
gathered_CD4Th17subset <- inner_join(metadata2, gathered_CD4Th17subset)
gathered_CD4Th17subset <- merge(as(gathered_CD4Th1subset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
# To resolve the duplicate gene_name column in this run, execute the following lines. Comment this section out if this error does not occur.
gathered_CD4Th17subset <- gathered_CD4Th17subset[,-1] # Drop the extra gene_nanes.y column
colnames(gathered_CD4Th17subset)[5] = "gene_name" # Rename the gene_name.x column to just gene_name

gathered_CD4Th17subset$normalized_counts <- as.numeric(as.character(gathered_CD4Th17subset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

# CD4Tfhsubset <- normalized_counts_CD4 %in%
#   filter(gene_name %in% Tfhgenes)
# gathered_CD4Tfhsubset <- CD4Tfhsubset %>%
#   gather(colnames(CD4Tfhsubset)[2:31], key = "sample_name", value="normalized_counts")
# gathered_CD4Tfhsubset <- inner_join(metadata2, gathered_CD4Tfhsubset)
# gathered_CD4Tfhsubset <- merge(as(gathered_CD4Tfhsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
# gathered_CD4Tfhsubset$normalized_counts <- as.numeric(as.character(gathered_CD4Tfhsubset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.
# 
# CD4Tregsubset <- normalized_counts_CD4 %in%
#   filter(gene_name %in% Treggenes)
# gathered_CD4Tregsubset <- CD4Tregsubset %>%
#   gather(colnames(CD4Tregsubset)[2:31], key = "sample_name", value="normalized_counts")
# gathered_CD4Tregsubset <- inner_join(metadata2, gathered_CD4Tregsubset)
# gathered_CD4Tregsubset <- merge(as(gathered_CD4Tregsubset,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Add column with gene names.
# gathered_CD4Tregsubset$normalized_counts <- as.numeric(as.character(gathered_CD4Tregsubset$normalized_counts)) # Ensures that the normalized count data are in the "numeric" class.

jitter <- position_jitter(width=0.1, height=0.1)
ggplot(gathered_precursorSubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Markers of Immaturity") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_PTCLsubgroups, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of GATA3 and TBX21") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_Activatedsubset, aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of CD25 and MHC II Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_CD4TCRsubset , aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Genes Associated with TCR Signaling") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_CD4Th1subset , aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th1 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_CD4Th2subset , aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th2 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

ggplot(gathered_CD4Th17subset , aes(x=gene_name, y=normalized_counts)) +
  geom_point(aes(color=phenotype), position=jitter) +
  scale_y_log10() +
  xlab("Genes") +
  ylab("log10 Normalized Counts") +
  ggtitle("Expression of Th17 Associated Genes") +
  theme_bw() +
  theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5))

# ggplot(gathered_CD4Tregsubset , aes(x=gene_name, y=normalized_counts)) +
#   geom_point(aes(color=phenotype), position=jitter) +
#   scale_y_log10() +
#   xlab("Genes") +
#   ylab("log10 Normalized Counts") +
#   ggtitle("Expression of Treg Associated Genes") +
#   theme_bw() +
#   theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
#   theme(plot.title = element_text(hjust=0.5))
# 
# ggplot(gathered_CD4Tfhsubset , aes(x=gene_name, y=normalized_counts)) +
#   geom_point(aes(color=phenotype), position=jitter) +
#   scale_y_log10() +
#   xlab("Genes") +
#   ylab("log10 Normalized Counts") +
#   ggtitle("Expression of Tfh Associated Genes") +
#   theme_bw() +
#   theme(axis.text.x = element_text(size=12, face="bold", angle=45, hjust=1)) +
#   theme(plot.title = element_text(hjust=0.5))
```


## 4.5 Correlation Matrices
Calculate the distances between each sample, then format the matrix.
```{r corr-matrix-setup, message=FALSE}
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists) #convert from data.frame -> matrix
rownames(sampleDistMatrix) <- paste(rld$phenotype, colnames(rld), sep="-") # Add sample labels
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #Pick colors
```

Draw the heatmap.
```{r corr-matrix, message=FALSE}
p <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

## 4.6 Volcano Plots
Volcano plots are a nice way of displaying the fold change against the p-value.
```{r EnhancedVolcano, message=FALSE, fig.height=4, fig.width=8.5, fig.fullwidth=TRUE}
library(EnhancedVolcano)
# Optional: Read in a DESeq2 results file with unnamed probe IDs removed. Otherwise, use DESeq2 results table for the comparison of interest.
volcano_data <- read.csv("../01_input/CD4PTCLvsCD4CTRL_DESeq2res_BlankIDsRemoved.csv")

EnhancedVolcano(volcano_data, 
                lab = volcano_data$gene_name,
                x = 'log2FoldChange',
                y = 'padj',
                title = 'CD4 PTCL vs. Control CD4 T-cells',
                FCcutoff = 1.5,
                pCutoff = 0.05,
                pointSize = 2.0,
                labSize = 4.0,
                labCol = 'black',
                labFace = 'bold',
                colAlpha = 4/5,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 6.0,
                drawConnectors = TRUE,
                widthConnectors = 0.2,
                colConnectors = 'black'
                )
```

```{r EnhancedVolcano-selectGenes, message=FALSE, fig.height=4.5, fig.width=6, fig.fullwidth=TRUE}
# Volcano plot with only key variables labeled:
EnhancedVolcano(volcano_data, 
                lab = volcano_data$gene_name,
                x = 'log2FoldChange',
                y = 'padj',
                selectLab = c('GATA3', 'PTEN', 'IFNG', 'EOMES', 'TBX21', 'KIT', 'DNTT', 'CD34', 'IL2RA', 'IL2RB', 'DLA-DRA', 'DLA-DQA1', 'HLA-DQB1', 'HLA-DQB2'),
                title = 'CD4 PTCL vs. CD4 CTRL',
                FCcutoff = 1.5,
                pCutoff = 0.05,
                pointSize = 2.0,
                labSize = 4.0,
                labCol = 'black',
                labFace = 'bold',
                colAlpha = 4/5,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 6.0,
                drawConnectors = TRUE,
                widthConnectors = 0.2,
                colConnectors = 'black'
                )
```

## 4.7 Hierarchically clustered heatmaps
### 4.7.1 Data Preparation

To make a heatmap featuring all groups (all PTCL subtypes and controls), subset the results table for just the differentially expressed genes.
```{r subset-significant-1, results='hide', message=FALSE}
sign_res <- subset(res, padj < 0.05)
```

Another method is to get the differentially expressed genes for each comparison, and combine them together.
```{r subset-significant-2, results='hide'}
sign_CD4PTCLvsCD4CTRL <- subset(CD4PTCLvsCD4CTRL_final_res, padj < 0.05)
sign_CD4PTCLvsCD8PTCL <- subset(CD4PTCLvsCD8PTCL_final_res, padj < 0.05)
sign_CD4PTCLvsDNPTCL <- subset(CD4PTCLvsDNPTCL_final_res, padj < 0.05)
sign_CD8PTCLvsCD8CTRL <- subset(CD8PTCLvsCD8CTRL_final_res, padj < 0.05)
sign_CD8PTCLvsDNPTCL <- subset(CD8PTCLvsDNPTCL_final_res, padj < 0.05)
sign_DNPTCLvsCD4CTRL <- subset(DNPTCLvsCD4CTRL_final_res, padj < 0.05)
sign_DNPTCLvsCD8CTRL <- subset(DNPTCLvsCD8CTRL_final_res, padj < 0.05)

# Merge
changing_genes <- rbind(sign_CD4PTCLvsCD4CTRL, sign_CD4PTCLvsCD8PTCL, sign_CD4PTCLvsDNPTCL, sign_CD8PTCLvsCD8CTRL, sign_CD8PTCLvsDNPTCL, sign_DNPTCLvsCD4CTRL, sign_DNPTCLvsCD8CTRL)
```

To make a heatmap of just two groups, subset just the differentially expressed genes of the results for that comparison, as generated earlier.
```{r subset-significant-3, results='hide'}
sign_CD4PTCLvsCD4CTRL <- subset(CD4PTCLvsCD4CTRL_final_res, padj < 0.05) #Subset the results table for just the differentially expressed genes.
```

For a heatmap that includes only the top 50 differentially expressed genes between CD4 PTCL vs CD4 Controls, subset the significantly differentially expressed genes table for the top 25 upregulated & downregulated genes.
```{r subset-top50, results='hide'}
top25up <- head(sign_CD4PTCLvsCD4CTRL[order(sign_CD4PTCLvsCD4CTRL$log2FoldChange, decreasing=T),], n = 25)
top25dn <- tail(sign_CD4PTCLvsCD4CTRL[order(sign_CD4PTCLvsCD4CTRL$log2FoldChange, decreasing=T),], n = 25)
top50genes <- rbind(top25up, top25dn) # Combine them
```

Use the r-stabilized log values of the data and subset just the significantly differentially expressed genes, or a specific gene list of interest. Use the assay function to convert this object to a data frame.

```{r heatmap-setup-1, results='hide'}
rld_matrix <- assay(rlog(dds, blind=FALSE))
```

```{r heatmap-setup-2, results='hide'}
signres_rld <- subset(rld_matrix, rownames(rld) %in% rownames(sign_res))
changing_genes_rld <- subset(rld_matrix, rownames(rld) %in% rownames(changing_genes))

# Subset CD4s only
CD4PTCLvsCD4CTRL_rld <- subset(rld_matrix, rownames(rld_matrix) %in% rownames(sign_CD4PTCLvsCD4CTRL), select = c("CF01", "CF02", "CF03", "CF05", "CF07", "CF12", "CF15", "CF17", "CF18", "CF19", "CF20", "CF22", "CF23", "CF24", "CF25", "CF26", "CF27", "CF28", "CF29", "CF30", "CF31", "CF32", "CF33", "CF34", "CF35", "CF36", "CF37", "CF39", "CF40"))

# Subset top 50
top50_rld <- subset(rld_matrix, rownames(rld_matrix) %in% rownames(top50genes), select = c("CF01", "CF02", "CF03", "CF05", "CF07", "CF12", "CF15", "CF17", "CF18", "CF19", "CF20", "CF22", "CF23", "CF24", "CF25", "CF26", "CF27", "CF28", "CF29", "CF30", "CF31", "CF32", "CF33", "CF34", "CF35", "CF36", "CF37", "CF39", "CF40"))

## Subset the gene list associated with the GATA3-PTCL and TBX21-PTCL subgroups in Iqbal et al.
Iqbal_genes <- c("GATA3", "CCR4", "IL18R1", "ACKR3", "IK", "MSH6", "EGR1", "CAT", "SEPTIN6", "TBX21", "EOMES", "CXCR3", "IL2RB", "CCL3", "IFNG") # A variable that contains the gene list we want included in the heatmap.

CD4_rld <- subset(rld_matrix, rownames(rld_matrix) %in% rownames(CD4PTCLvsCD4CTRL_final_res), select = c("CF01", "CF02", "CF03", "CF05", "CF07", "CF12", "CF15", "CF17", "CF18", "CF19", "CF20", "CF22", "CF23", "CF24", "CF25", "CF26", "CF27", "CF28", "CF29", "CF30", "CF31", "CF32", "CF33", "CF34", "CF35", "CF36", "CF37", "CF39", "CF40")) # Subset the rld_matrix for just the CD4 PTCLs and CD4 CTRLs.

# Add a gene name column to the rld files.
Iqbal_rld <- as.data.frame(CD4_rld, drop=FALSE) # Convert the r-stabilized logs of the CD4 PTCL vs CD4 CTRL comparison to a data frame.
Iqbal_rld$probe_id <- rownames(Iqbal_rld) # Add probe ID as a column
Iqbal_rld <- merge(as(Iqbal_rld,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds a column of gene names for the associated probe ID

# Subset for just the gene list of interest.
Iqbal_subset <- Iqbal_rld %>%
  filter(gene_name %in% Iqbal_genes)

# Clean up the matrix for the heatmap. We want only sample IDs as columns and gene symbols as rownames.
rownames(Iqbal_subset) <- Iqbal_subset$gene_name #Make the gene symbols the rownames.
# colnames(Iqbal_subset) # See which columns have the probe ID and the gene symbol. (Comment this out before knitting document.)
Iqbal_subset <- Iqbal_subset[, 2:30] # Subset just the patient ID columns.

## Subset only the rlog normalized count data for tumor samples.
tumors <- c("CF01", "CF02",	"CF03", "CF04", "CF05",	"CF06",	"CF07",	"CF08",	"CF09",	"CF10",	"CF11",	"CF15",	"CF18",	"CF19",	"CF20",	"CF22",	"CF23",	"CF24",	"CF25",	"CF26",	"CF27",	"CF28",	"CF31",	"CF32",	"CF33",	"CF34",	"CF35",	"CF36",	"CF37",	"CF38",	"CF39",	"CF40")
All_tumors_rld <- subset(rld_matrix, rownames(rld_matrix) %in% rownames(rld_matrix), select = tumors)
All_tumors <- as.data.frame(All_tumors_rld, drop=FALSE) # Convert rlogs to a data frame
All_tumors$probe_id <- rownames(All_tumors) # Add probe ID as a column
All_tumors <- merge(as(All_tumors,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds column of gene names for the associated probe ID
Iqbal_tumors <- All_tumors %>%
  filter(gene_name %in% Iqbal_genes) # Subset for just the gene list of interest.
rownames(Iqbal_tumors) <- Iqbal_tumors$gene_name # Make the gene symbols the rownames
Iqbal_tumors <- Iqbal_tumors[, 2:33] # Subset just the patient ID columns.

## Subset the Iqbal gene set for all samples.
rld_df <- as.data.frame(rld_matrix, drop=FALSE) # Convert the r-stabilized logs of all samples to a data frame.
rld_df$probe_id <- rownames(rld_df) # Add probe Id column
head(rld_df)
rld_df <- merge(as(rld_df,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds column of gene names for the associated probe ID
Iqbal_all <- rld_df %>%
  filter(gene_name %in% Iqbal_genes) # Subset for just the gene list of interest.
rownames(Iqbal_all) <- Iqbal_all$gene_name # Make the gene symbols the rownames
dim(Iqbal_all)
head(Iqbal_all)
Iqbal_all <- Iqbal_all[, 2:40] # Subset just the patient ID columns.
```

Before proceeding to draw the heatmap, we need metadata files to allow annotation of phenotype on the heatmap.
```{r heatmap-setup-3, results='hide'}
# For the heatmap of all phenotypes:
metadata_all <- metadata1[,4, drop=FALSE] # This pulls just the phenotype column out of our metadata file and keeps the rownames, which are our patient IDs.

# For the CD4 PTCL v CD4 CTRL heatmap, it is easier to import a metadata file that only has information for the CD4 PTCL and CD4 CTRL samples.
samplePhenotype <- read.csv("../01_input/metadata_CD4sOnly_outlierRemoved.csv")
rownames(samplePhenotype) <- colnames(CD4PTCLvsCD4CTRL_rld)
samplePhenotype <- samplePhenotype[,2, drop=FALSE] # Since we don't want the sample names annotated, remove this column now that we have rownames assigned to the appropriate sample ID.
```

We will use the genenames metadata file to annotate the heatmap of the top 50 differentially expressed genes. Extract from the metadata file only those genes which are represented in the list of the top 50 differentially expressed.
```{r heatmap-setup-4, results='hide'}
rownames(genenames) <- genenames$probe_id # Assign the probe ID to the row name of the metadata file
genenames2 <- genenames[,2, drop=FALSE] # Remove the column of probe IDs now that we've assigned them as rownames.
top50_genenames <- subset(genenames2, rownames(genenames2) %in% rownames(top50genes)) # Extract only those in the top 50
```

### Draw the heatmaps.
#### Heatmap of all phenotypes, from subsetting the significant genes in 'res':
```{r pheatmap-all}
p <- pheatmap(signres_rld, 
              scale="row", 
              color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
              cluster_rows=TRUE, 
              cluster_cols=TRUE, 
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "complete",
              annotation_col = metadata_all,
              show_rownames = FALSE)
```

#### Heatmap of all phenotypes, from subsetting the significant genes by rbinding all comparisons together:
```{r pheatmap-rbindChangingGenes}
p <- pheatmap(changing_genes_rld, 
              scale="row", 
              color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
              cluster_rows=TRUE, 
              cluster_cols=TRUE, 
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "complete",
              annotation_col = metadata_all,
              show_rownames = FALSE)
```

#### Heatmap of CD4 PTCL vs CD4 Control samples only:
```{r pheatmap-CD4sOnly}
p <- pheatmap(CD4PTCLvsCD4CTRL_rld, 
         scale="row", 
         color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
         cluster_rows=TRUE, 
         cluster_cols=TRUE, 
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         annotation_col = samplePhenotype,
         show_rownames = FALSE)
```

#### Heatmap of just the top 50 differentially expressed genes in the CD4 PTCL vs CD4 Control analysis:
```{r heatmap-top50DEG}
annot_colors = list(Phenotype=c(CD4_CTRL="#32e3e3", CD4_PTCL="#FF9999")) # Customizing the colors of the phenotype labels.
ptop50 <- pheatmap(top50_rld, 
              scale="row", 
              color = colorRampPalette(c("blue", "white", "red"), space = "Lab")(100),
              cluster_rows=TRUE,
              cluster_cols=TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "complete",
              annotation_col = samplePhenotype,
              annotation_row = top50_genenames,
              annotation_colors = annot_colors,
              show_rownames = FALSE)
```

# 5. Final data export

## 5.1 Export the matrices of rlog transformed normalized counts:
```{r}
write.csv(Iqbal_subset, file="../03_output/Rlog_Transformed_Matrix_of_Normalized_Values_IqbalGeneSubset.csv")
write.csv(Iqbal_tumors, file="../03_output/Rlog_Transformed_Matrix_IqbalGenes_AllTumors.csv")
write.csv(Iqbal_all, file="../03_output/Rlog_Transformed_Matrix_IqbalGenes")

CD4rld_matrix <- as.data.frame(CD4PTCLvsCD4CTRL_rld, drop=FALSE) # Convert the r-stabilized logs to a data frame.
CD4rld_matrix$probe_id <- rownames(CD4rld_matrix) # Add probe ID as a column
CD4rld_matrix <- merge(as(CD4rld_matrix,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds a column of gene names for the associated probe ID
write.csv(CD4rld_matrix, file="../03_output/Rlog_Transformed_Matrix_of_Normalized_Values_CD4sOnly.csv")

rld_matrix <- as.data.frame(rld_matrix, drop=FALSE) # Convert the r-stabilized logs to a data frame.
rld_matrix$probe_id <- rownames(rld_matrix) # Add probe ID as a column
rld_matrix <- merge(as(rld_matrix,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds a column of gene names for the associated probe ID
write.csv(rld_matrix, file="../03_output/Rlog_Transformed_Matrix_of_NormalizedCounts_AllSamples.csv")
```

## 5.2 Capture and export the loadings used for PCA:
```{r, results='hide'}
# Capture the loadings for PCA:
rv <- rowVars(assay(rld))
select <- order(rv, decreasing=TRUE)[seq_len(min(500, length(rv)))]
pca <- prcomp(t(assay(rld)[select,]))
loadings <- as.data.frame(pca$rotation)
loadings$probe_id <- rownames(loadings) # Add probe ID column
head(loadings)
head(genenames)
loadings <- merge(as(loadings,"data.frame"), genenames, by="probe_id", all.x=TRUE) # Adds a column of gene names for the associated probe ID
write.csv(loadings, file="220922_PCA_loadings_withGeneNames.csv") # Export to .csv
```

# 6. Citations
```{r}
sessionInfo()
citation()
```

